app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoViewModel.kt

```diff
@@ -0,0 +1,329 @@
+package io.zenandroid.onlinego.ui.screens.puzzle
+
+import android.content.Context
+import android.util.Log
+import android.widget.Toast
+import androidx.compose.runtime.State
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import io.zenandroid.onlinego.data.model.Cell
+import io.zenandroid.onlinego.data.model.Mark
+import io.zenandroid.onlinego.data.model.StoneType
+import io.zenandroid.onlinego.data.model.ogs.MoveTree
+import io.zenandroid.onlinego.data.model.ogs.PlayCategory
+import io.zenandroid.onlinego.data.model.ogs.PuzzleRating
+import io.zenandroid.onlinego.data.model.ogs.PuzzleSolution
+import io.zenandroid.onlinego.data.model.local.Puzzle
+import io.zenandroid.onlinego.data.repositories.PuzzleRepository
+import io.zenandroid.onlinego.gamelogic.RulesManager
+import io.zenandroid.onlinego.gamelogic.Util
+import io.zenandroid.onlinego.gamelogic.Util.toCoordinateSet
+import io.zenandroid.onlinego.mvi.MviView
+import io.zenandroid.onlinego.mvi.Store
+import io.zenandroid.onlinego.utils.addToDisposable
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import org.koin.core.context.GlobalContext
+import java.time.Instant.now
+import java.time.temporal.ChronoUnit.*
+
+class TsumegoViewModel (
+    private val puzzleRepository: PuzzleRepository,
+    private val store: Store<TsumegoState, TsumegoAction>,
+    private val puzzleId: Long
+): ViewModel() {
+    private val _state = MutableStateFlow(TsumegoState())
+    val state: StateFlow<TsumegoState> = _state
+    private val subscriptions = CompositeDisposable()
+    var collectionPuzzles by mutableStateOf(emptyList<Puzzle>())
+        private set
+
+    private val wiring = store.wire()
+    private var viewBinding: Disposable? = null
+
+    private var cursor by mutableStateOf(0)
+
+    private var moveReplyJob: Job? = null
+
+    override fun onCleared() {
+        wiring.dispose()
+    }
+
+    init {
+        puzzleRepository.getPuzzle(puzzleId)
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(this::setPuzzle, this::onError)
+            .addToDisposable(subscriptions)
+        fetchRating(puzzleId)
+    }
+
+    private fun setPuzzle(puzzle: Puzzle, attempts: Int = 1) {
+        _state.update {
```

Comment:
this might actually be better than the method I was using `_state.value = _state.value.copy(...)` as it's atomic. I will make a mental note to use it myself.